### PR TITLE
Only retry failed messages instead the whole batch

### PR DESCRIPTION
**Pull request by xavi-**
Resubmitting the whole batch probably resulted in duplicate data

Context: https://clever.atlassian.net/browse/INFRA-2038

### CODE REVIEW

**Comment by jiancheung** [comment]
Do we have tests for this/How has this been tested?

**Comment by xavi-** [comment]
Typically with heka-testing but docker is having issues at the moment.  Also with dangerous logging changes we tend to test changes in dev for a day or so.

File: aws/firehose.go [hunk 1]
```diff
@@ -56,13 +59,57 @@ func (f Firehose) PutRecordBatch(records [][]byte) error {
...
+	}
+
+	// Expotential backoff with retry limit
+	return f.retryFailedRecords(retryRecords, newRes, retries-1, delay*2)
+}
+
+// PutRecordBatch sends an array of records to the Firehose stream
+// as a single batch request
+func (f Firehose) PutRecordBatch(records [][]byte) error {
+	res, err := f.sendRecords(records)
```
Comment (xavi-) [inline aws/firehose.go]: I feel that combining `sendRecords` and `retryFailedRecords` would make things more confusing.  I think the root issue is that you need a `PutRecordBatchOutput` struct on the first iteration.  I could make `retryFailedRecords` non-recursive but then there'd be lines that looks like this:

```golang
records = retryRecords
res = newRes
```

Which I thought would be confusing.  Just calling `f.retryFailedRecords` again seemed simpler.

Comment (xavi-) [inline aws/firehose.go]: Merged in your changes

File: aws/firehose.go [hunk 2]
```diff
@@ -56,13 +59,47 @@ func (f Firehose) PutRecordBatch(records [][]byte) error {
...
 	if err != nil {
 		return err
 	}
-	// Check for any individual records failing
-	if *res.FailedPutCount != 0 {
-		return fmt.Errorf("%d records failed to upload", *res.FailedPutCount)
+
+	retries := 0
+	delay := 250
+	for *res.FailedPutCount != 0 {
```
Comment (jiancheung) [inline aws/firehose.go]: it could also be 
```
for err == nil && *res.FailedPutcount != 0 {
...
}
return err
```
But I don't think I like that either.

**Comment by jiancheung** [comment]
Assuming that you verified that code actually runs and works, lgtm!

File: aws/firehose.go (hunk 2 above)
Comment (xavi-) [inline aws/firehose.go]: I love `for *res.FailedPutCount != 0 {`

**Comment by xavi-** [comment]
Got it working in heka-testing, gonna merge and deploy into dev

**Pull request merged**